src/Module.Server/Rewards/CrpgRewardServer.cs

```diff
@@ -195,17 +210,78 @@ public override void OnRemoveBehavior()
             return;
         }
 
+        CalculatePlayerUpkeepCompensations();
+        CompensateTeamhitPlayers(userUpdates);
+
         // TODO: add retry mechanism (the endpoint need to be idempotent though).
         try
         {
-            var res = (await _crpgClient.UpdateUsersAsync(new CrpgGameUsersUpdateRequest { Updates = userUpdates })).Data!;
+            var res = (await _crpgClient.UpdateUsersAsync(new CrpgGameUsersUpdateRequest { Updates = userUpdates.Values.ToList() })).Data!;
             SendRewardToPeers(res.UpdateResults, crpgPeerByUserId, valorousPlayerIds);
         }
         catch (Exception e)
         {
             Debug.Print("Couldn't update users: " + e);
             SendErrorToPeers(crpgPeerByUserId);
         }
+
+        _compensationTable.Clear();
+    }
+
+    private void CalculatePlayerUpkeepCompensations()
+    {
+        foreach (var affectedEntry in _agentsThatGotTeamhitByCrpgPeer)
+        {
+            Agent affectedAgent = affectedEntry.Value;
+            CrpgPeer affectedCrpgPeer = affectedEntry.Key;
+            foreach (Agent.Hitter affector in affectedAgent.HitterList)
+            {
+                CrpgPeer affectorCrpgPeer = affector.HitterPeer.GetNetworkPeer().GetComponent<CrpgPeer>();
+                if (affectedCrpgPeer == affectorCrpgPeer)
+                {
+                    continue;
+                }
+
+                Console.WriteLine($"" + affector.HitterPeer.Name + " did " + affector.Damage + " Total damage to " + affectedAgent.Name + " total life was " + affectedAgent.BaseHealthLimit);
+                Console.WriteLine($"The compensation% would be " + (affector.Damage / affectedAgent.BaseHealthLimit * 100));
+
+                float compensationRatio = affector.Damage / affectedAgent.BaseHealthLimit;
+                int compensatedRepairCost = (int)Math.Floor(_compensationTable[affectedCrpgPeer].TotalRepairCost * compensationRatio);
+
+                _compensationTable[affectedCrpgPeer].CompensatedRepairCost += compensatedRepairCost;
+                _compensationTable[affectorCrpgPeer].RepairCostToCompensate += compensatedRepairCost;
+            }
+        }
+
+        _agentsThatGotTeamhitByCrpgPeer.Clear();
+    }
+
+    private void CompensateTeamhitPlayers(Dictionary<CrpgPeer, CrpgUserUpdate> userUpdates)
+    {
+        foreach (var entry in _compensationTable)
+        {
+            CrpgPeer crpgPeer = entry.Key;
+            CrpgUserUpkeepCompensation crpgUserUpkeepCompensation = entry.Value;
+
+            if (crpgUserUpkeepCompensation.CompensatedRepairCost == 0 && crpgUserUpkeepCompensation.RepairCostToCompensate == 0)
+            {
+                continue;
+            }
+
+            CrpgUserReward? reward = userUpdates[crpgPeer].Reward;
+            if (reward == null)
+            {
+                continue;
+            }
+

```

Comment:
Correct me if i'm wrong , but you're still hacking the reward?
Consider using userUpdates.Add(userUpdate); , and adding a new reward just for gold compensation. Will help decouple compensation from upkeep and gold reward

Reply:
Done